app/scripts/lib/rpc-method-middleware/handlers/wallet-requestPermissions.test.ts

```diff
@@ -0,0 +1,756 @@
+        updateCaveat,
+        grantPermissions,
+        requestPermissionApprovalForOrigin,
+        getAccounts,
+      },
+    );
+
+  return {
+    response,
+    next,
+    end,
+    requestPermissionsForOrigin,
+    getPermissionsForOrigin,
+    updateCaveat,
+    grantPermissions,
+    requestPermissionApprovalForOrigin,
+    getAccounts,
+    handler,
+  };
+};
+
+describe('requestPermissionsHandler', () => {
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  beforeEach(() => {
+    MockMultichain.setEthAccounts.mockImplementation(
+      (caveatValue) => caveatValue,
+    );
+    MockMultichain.setPermittedEthChainIds.mockImplementation(
+      (caveatValue) => caveatValue,
+    );
+  });
+
+  it('returns an error if params is malformed', async () => {
+    const { handler, end } = createMockedHandler();
+
+    const malformedRequest = {
+      ...getBaseRequest(),
+      params: [],
+    };
+    await handler(malformedRequest);
+    expect(end).toHaveBeenCalledWith(
+      invalidParams({ data: { request: malformedRequest } }),
+    );
+  });
+
+  it('requests approval from the ApprovalController for eth_accounts and permittedChains when only eth_accounts is specified in params and origin is not snapId', async () => {
+    const { handler, requestPermissionApprovalForOrigin } =
+      createMockedHandler();
+
+    await handler({
+      ...getBaseRequest(),
+      params: [
+        {
+          [RestrictedMethods.eth_accounts]: {
+            foo: 'bar',
+          },
+        },
+      ],
+    });
+
+    expect(requestPermissionApprovalForOrigin).toHaveBeenCalledWith({
+      [RestrictedMethods.eth_accounts]: {
+        foo: 'bar',
+      },
+      [PermissionNames.permittedChains]: {},
+    });
+  });
+
+  it('requests approval from the ApprovalController for eth_accounts and permittedChains when only permittedChains is specified in params and origin is not snapId', async () => {
+    const { handler, requestPermissionApprovalForOrigin } =
+      createMockedHandler();
+
+    await handler({
+      ...getBaseRequest(),
+      params: [
+        {
+          [PermissionNames.permittedChains]: {
+            caveats: [
+              {
+                type: CaveatTypes.restrictNetworkSwitching,
+                value: ['0x64'],
+              },
+            ],
+          },
+        },
+      ],
+    });
+
+    expect(requestPermissionApprovalForOrigin).toHaveBeenCalledWith({
+      [RestrictedMethods.eth_accounts]: {},
+      [PermissionNames.permittedChains]: {
+        caveats: [
+          {
+            type: CaveatTypes.restrictNetworkSwitching,
+            value: ['0x64'],
+          },
+        ],
+      },
+    });
+  });
+
+  it('requests approval from the ApprovalController for eth_accounts and permittedChains when both are specified in params and origin is not snapId', async () => {
+    const { handler, requestPermissionApprovalForOrigin } =
+      createMockedHandler();
+
+    await handler({
```

Comment:
Similar for this one too.

Reply:
https://github.com/MetaMask/metamask-extension/pull/27847/commits/f7a09fda54e6e4998b098b2ea84e8557b0e13d3d